target/ppc/translate/fixedpoint-impl.c.inc

```diff
@@ -492,3 +492,32 @@ static bool trans_PEXTD(DisasContext *ctx, arg_X *a)
 #endif
     return true;
 }
+
+static bool do_hash(DisasContext *ctx, arg_X *a, bool priv,
+    void (*helper)(TCGv_ptr, TCGv, TCGv, TCGv))
+{
+    TCGv ea = tcg_temp_new();
+
+    if (!(ctx->insns_flags2 & PPC2_ISA310)) {
+        /* if version is before v3.1, this operation is a nop */
+        return true;
+    }
+
+    if (unlikely(priv && ctx->pr)) {
+        /* if instruction is privileged but the context is in user space */
+        gen_priv_exception(ctx, POWERPC_EXCP_PRIV_OPC);
+        return 0;
+    }
```

Comment:
Maybe with
```suggestion
    if (priv) {
        CHK_SV;
        return true;
    }
```
`hashchkp` and `hashstp` could be optimized to just v3.1 check and `GEN_PRIV` on linux-user.

Reply:
Wouldn't this cause a situation where priv is true, but `ctx->pr == false` and it `return true` anyway instead of continuing?

Reply:
And also, CHK_SV returns void, which doesn't work here

Reply:
Hum, I guess we need a decodetree-era version of `CHK_*`/`GEN_PRIV`, returning a boolean instead of void. This feels a bit like a mix of `REQUIRE_64BIT` and  `REQUIRE_{VECTOR,VSX,FPU}` 
```c
#if defined(CONFIG_USER_ONLY)
# define REQUIRE_SV_PRIV(CTX)                               \
    do {                                                    \
        gen_priv_exception(CTX, POWERPC_EXCP_PRIV_OPC);     \
        return true;                                        \
    } while (0)
#else
# define REQUIRE_SV_PRIV(CTX)                               \
    do {                                                    \
        if (unlikely(ctx->pr)) {                            \
            gen_priv_exception(CTX, POWERPC_EXCP_PRIV_OPC); \
            return true;                                    \
        }                                                   \
    } while (0)
#endif
```

Or maybe with a method instead of a macro
```c
bool do_check_sv(DisasContext *ctx)
{
#if !defined(CONFIG_USER_ONLY)
    if (likely(ctx->pr != 0)) {
        return false;
    }
#endif
    gen_priv_exception(ctx, POWERPC_EXCP_PRIV_OPC);
    return true
}
```
and then in the instruction code
```c
static bool do_hash(DisasContext *ctx, arg_X *a, bool priv,
    void (*helper)(TCGv_ptr, TCGv, TCGv, TCGv))
{
    if (do_check_sv(ctx)) {
        return true;
    }

    /* ... */
}
```
@lrcoutinho will probably benefit from this in his `slb*` endeavor.

Reply:
What do you think about [449b18b](https://github.com/PPC64/qemu/pull/61/commits/449b18b42a7f54525462756e89097ca8426d55cd)?

Reply:
Looks good, but drop the `inline`, the compiler should know better.